core/src/main/java/com/datastrato/gravitino/storage/relational/service/CatalogMetaService.java

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * The service class for catalog metadata. It provides the basic database operations for catalog.
+ */
+public class CatalogMetaService {
+  private static final CatalogMetaService INSTANCE = new CatalogMetaService();
+
+  public static CatalogMetaService getInstance() {
+    return INSTANCE;
+  }
+
+  private CatalogMetaService() {}
+
+  public CatalogEntity getCatalogByIdentifier(NameIdentifier identifier) {
+    checkCatalogNamespaceWithIdentifier(identifier);
+    String metalakeName = identifier.namespace().level(0);
+    String catalogName = identifier.name();
+    Long metalakeId =
+        SessionUtils.getWithoutCommit(
+            MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+    if (metalakeId == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, identifier.namespace().toString());
+    }
+    CatalogPO catalogPO =
+        SessionUtils.getWithoutCommit(
+            CatalogMetaMapper.class,
+            mapper -> mapper.selectCatalogMetaByMetalakeIdAndName(metalakeId, catalogName));
+    if (catalogPO == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, identifier.toString());
+    }
+    return POConverters.fromCatalogPO(catalogPO, identifier.namespace());
+  }
+
+  public List<CatalogEntity> listCatalogsByNamespace(Namespace namespace) {
+    checkCatalogNamespace(namespace);
+    String metalakeName = namespace.level(0);
+    Long metalakeId =
+        SessionUtils.getWithoutCommit(
+            MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+    if (metalakeId == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, namespace.toString());
+    }
+    List<CatalogPO> catalogPOS =
+        SessionUtils.getWithoutCommit(
+            CatalogMetaMapper.class, mapper -> mapper.listCatalogPOsByMetalakeId(metalakeId));
+    return POConverters.fromCatalogPOs(catalogPOS, namespace);
+  }
+
+  public void insertCatalog(CatalogEntity catalogEntity, boolean overwrite) {
+    try {
+      checkCatalogNamespaceWithIdentifier(catalogEntity.nameIdentifier());
+      Long metalakeId =
+          SessionUtils.getWithoutCommit(
+              MetalakeMetaMapper.class,
+              mapper -> mapper.selectMetalakeIdMetaByName(catalogEntity.namespace().level(0)));
+      if (metalakeId == null) {
+        throw new NoSuchEntityException(
+            NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, catalogEntity.namespace().toString());
+      }
+      SessionUtils.doWithCommit(
+          CatalogMetaMapper.class,
+          mapper -> {
+            CatalogPO po = POConverters.initializeCatalogPOWithVersion(catalogEntity, metalakeId);
+            if (overwrite) {
+              mapper.insertCatalogMetaOnDuplicateKeyUpdate(po);
+            } else {
+              mapper.insertCatalogMeta(po);
```

Comment:
Can we use `exists` and  judge `EntityAlreadyExistsException` here?

Reply:
In the original metalake implementation, I referred to the implementation of kv here, but after discussing with @qqqttt123 , I changed it to capture the database unique constraint exception to determine whether it exists. I have no special preference for this, but we need to reach an agreement.

Reply:
OK, cc @qqqttt123

Reply:
The operations exists and insert aren't a transaction. So although we judge the exist, the operation of the insert may still occur exist exception.

Reply:
We have added a global lock to prevent concurrent modification, deletion, and creation operations, so I think we can simplify it.

Reply:
@yuqi1129 I paid attention to the implementation of global locks. Is it currently a JVM-level lock implementation, right? If there are multiple server nodes, is it still possible to have parallel inserts and updates?

Reply:
> Is it currently a JVM-level lock implementation, right? 

Yes. 

> If there are multiple server nodes, is it still possible to have concurrent inserts and updates?

We cannot guarantee simultaneous modification, insertion, and creation in multiple server instances.

Reply:
After using MySQL as the backend, we may deploy multiple server nodes to accept the requests, so if we want to maintain a global lock, it may need to be a distributed lock

Reply:
If we want to maintain the transactions, it may be a simpler solution to leave it to MySQL

Reply:
> If we want to maintain the transactions, it may be a simpler solution to leave it to MySQL

Regarding transactions, we cannot rely on the underlying storage to provide a transaction mechanism. What if some storage can't support it?

Reply:
Maybe in the future we need to clearly point out which supported storage can only deploy servers on a single node, and which storage can deploy servers on multiple nodes. But for now, global locks are not a problem for MySQL Backend. Even if it is not a distributed lock, it will not cause problems with concurrent insertion or updates when deploying servers on multiple nodes.

Reply:
Total agree.